staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go

```diff
@@ -0,0 +1,264 @@
+	// Consequently the filter can load a `*RMState` and work with it
+	// without concern for concurrent updates.  When a priority level is
+	// finally deleted, this will also involve storing a new `*RMState`
+	// pointer here, but in this case the range of the
+	// `RMState.priorityLevels` will be reduced --- by removal of the
+	// priority level that is no longer in use.
+	curState atomic.Value
+}
+
+// TypedConfigObjectReference is a reference to a relevant config API object.
+// No namespace is needed because none of these objects is namespaced.
+type TypedConfigObjectReference struct {
+	Kind string
+	Name string
+}
+
+func (tr *TypedConfigObjectReference) String() string {
+	return tr.Kind + "/" + tr.Name
+}
+
+// rmSetup is invoked at startup to create the infrastructure of this filter
+func rmSetup(loopbackClientConfig *restclient.Config, serverConcurrencyLimit int, requestWaitLimit time.Duration) *requestManagement {
+	reqMgmt := &requestManagement{
+		configQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "req_mgmt_config_queue"),
+		serverConcurrencyLimit: serverConcurrencyLimit,
+		requestWaitLimit:       requestWaitLimit,
+	}
+	// TODO: finish implementation
+	return reqMgmt
+}
+
+// WithRequestManagement limits the number of in-flight requests in a fine-grained way
+func WithRequestManagement(
+	handler http.Handler,
+	loopbackClientConfig *restclient.Config,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+) http.Handler {
+	reqMgmt := rmSetup(loopbackClientConfig, serverConcurrencyLimit, requestWaitLimit)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no RequestInfo found in context, handler chain must be wrong"))
+			return
+		}
+
+		// Skip tracking long running events.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		rmState := reqMgmt.curState.Load().(*RMState)
+		fs := reqMgmt.pickFlowSchema(r, rmState.flowSchemas, rmState.priorityLevelStates)
+		ps := reqMgmt.requestPriorityState(r, fs, rmState.priorityLevelStates)
+		if ps.config.Exempt {
+			klog.V(5).Infof("Serving %v without delay\n", r)
+			handler.ServeHTTP(w, r)
+			return
+		}
+		flowDistinguisher := reqMgmt.computeFlowDistinguisher(r, fs)
+		hashValue := reqMgmt.hashFlowID(fs.Name, flowDistinguisher)
+		execute, afterExecute := ps.fqs.Wait(hashValue, ps.config.HandSize)
+		if execute {
+			klog.V(5).Infof("Serving %v after queuing\n", r)
+			timedOut := ctx.Done()
+			finished := make(chan struct{})
+			go func() {
+				handler.ServeHTTP(w, r)
+				close(finished)
+			}()
+			select {
+			case <-timedOut:
```

Comment:
let's strip the timeout from the system, another filter will deal with it

Reply:
Yes, the upstream timeout filter will return after the timeout expires.  But in this filter we want an upper bound on execution time; my design analyses were predicated on that.  The logic here ensures that there is such an upper bound, as far as this filter is concerned.